precompiles/parachain-staking/src/lib.rs

```diff
@@ -232,286 +126,392 @@ where
 
+		// Build output.
 		Ok(PrecompileOutput {
 			exit_status: ExitSucceed::Returned,
-			cost: gas_consumed,
-			output: bool_to_solidity_bytes(is_candidate),
-			logs: Default::default(),
+			cost: gasometer.used_gas(),
+			output: EvmDataWriter::new().write(points).build(),
+			logs: vec![],
 		})
 	}
 
-	fn is_selected_candidate(input: &[u8]) -> Result<PrecompileOutput, ExitError> {
-		// parse the address
-		let candidate = H160::from_slice(&input[12..32]);
-
-		log::trace!(
-			target: "staking-precompile",
-			"Checking whether {:?} is a selected collator",
-			candidate
-		);
+	// Role Verifiers
 
-		// fetch data from pallet
-		let is_selected =
-			parachain_staking::Pallet::<Runtime>::is_selected_candidate(&candidate.into());
+	fn is_nominator(
+		mut input: EvmDataReader,
+		target_gas: Option<u64>,
+	) -> Result<PrecompileOutput, ExitError> {
+		let mut gasometer = Gasometer::new(target_gas);
 
-		log::trace!(target: "staking-precompile", "Result from pallet is {:?}", is_selected);
+		// Read input.
+		input.expect_arguments(1)?;
+		let raw_address: H160 = input.read::<Address>()?.into();
+		let address: Runtime::AccountId = raw_address.into();
 
-		let gas_consumed = <Runtime as pallet_evm::Config>::GasWeightMapping::weight_to_gas(
-			<Runtime as frame_system::Config>::DbWeight::get().read,
-		);
+		// Fetch info.
+		gasometer.record_cost(RuntimeHelper::<Runtime>::db_read_gas_cost())?;
+		let is_nominator = parachain_staking::Pallet::<Runtime>::is_nominator(&address);
 
+		// Build output.
 		Ok(PrecompileOutput {
 			exit_status: ExitSucceed::Returned,
-			cost: gas_consumed,
-			output: bool_to_solidity_bytes(is_selected),
-			logs: Default::default(),
+			cost: gasometer.used_gas(),
+			output: EvmDataWriter::new().write(is_nominator).build(),
+			logs: vec![],
 		})
 	}
 
-	fn min_nomination() -> Result<PrecompileOutput, ExitError> {
-		// fetch data from pallet
-		let raw_min_nomination: u128 = <
-			<Runtime as parachain_staking::Config>::MinNomination
-				as Get<BalanceOf<Runtime>>
-			>::get().try_into()
-				.map_err(|_|
-					error("Amount is too large for provided balance type")
-				)?;
-		let min_nomination: U256 = raw_min_nomination.into();
-
-		log::trace!(target: "staking-precompile", "Result from pallet is {:?}", min_nomination);
-
-		// TODO find cost of Config associated type read
-		// For now assume it is as bad as a storage read in the worst case
-		let gas_consumed = <Runtime as pallet_evm::Config>::GasWeightMapping::weight_to_gas(
-			<Runtime as frame_system::Config>::DbWeight::get().read,
-		);
+	fn is_candidate(
+		mut input: EvmDataReader,
+		target_gas: Option<u64>,
+	) -> Result<PrecompileOutput, ExitError> {
+		let mut gasometer = Gasometer::new(target_gas);
 
-		let mut buffer = [0u8; 32];
-		min_nomination.to_big_endian(&mut buffer);
+		// Read input.
+		input.expect_arguments(1)?;
+		let raw_address: H160 = input.read::<Address>()?.into();
+		let address: Runtime::AccountId = raw_address.into();
```

Comment:
This could be less verbose -- I added a `fn read_address<AccountId: From<H160>>` to precompile-utils [in 632](https://github.com/PureStake/moonbeam/pull/632/commits/9994b37fb73c6079bc5aed47a74c3ce57be134ce) which would improve this.

I'm considering adding it again in this PR unless @librelois or @nanocryk know of a less verbose way of expressing `input.read` with the existing code such that it returns `Runtime::AccountId`.

Reply:
There is more than one instance of this code which is why I'm proposing something less verbose.

Reply:
Instead, you case use:

```
let address: Runtime::AccountId = input.read::<Address>()?.0.into();
```

Reply:
Thank you @librelois !